modules/core/core-elasticsearch/src/main/java/com/enonic/xp/elasticsearch/impl/ElasticsearchClientActivator.java

```diff
@@ -0,0 +1,182 @@
+    private ServiceRegistration<EsClient> clientReg;
+
+    private BundleContext context;
+
+    private ScheduledExecutorService activateExecutorService;
+
+    private volatile boolean isRegistered = false;
+
+    @Activate
+    @SuppressWarnings("WeakerAccess")
+    public void activate( final BundleContext context )
+        throws InterruptedException
+    {
+        this.context = context;
+        this.client = new EsClient( "localhost", 9200 );
+
+        doActivate();
+    }
+
+    private void doActivate()
+        throws InterruptedException
+    {
+        final CountDownLatch activateLatch = new CountDownLatch( 1 );
+        this.activateExecutorService = Executors.newSingleThreadScheduledExecutor();
+
+        try
+        {
+            activateExecutorService.scheduleWithFixedDelay( () -> {
+                try
+                {
+                    if ( !isRegistered )
+                    {
+                        doRegisterElasticsearchClient();
+                    }
+                    else
+                    {
+                        doMonitorElasticsearchHealth();
+                    }
+                }
+                catch ( Exception e )
+                {
+                    handleException( e );
+                }
+                finally
+                {
+                    activateLatch.countDown();
+                }
+            }, 0, 1, TimeUnit.SECONDS );
+
+            activateLatch.await();
+        }
+        catch ( Exception e )
+        {
+            activateExecutorService.shutdown();
+            throw e;
+        }
+    }
+
+    private void doRegisterElasticsearchClient()
+    {
+        client.ping();
+
+        final ClusterHealthResponse healthResponse = client.clusterHealth( new ClusterHealthRequest() );
+
+        if ( healthResponse.getStatus() != ClusterHealthStatus.RED )
+        {
+            register();
+
+            LOG.info( "Elasticsearch is up." );
+        }
+    }
+
+    private void doMonitorElasticsearchHealth()
+    {
+        try
+        {
+            boolean pingSucceeded = client.ping();
+
+            final ClusterHealthResponse healthResponse = client.clusterHealth( new ClusterHealthRequest() );
+
+            if ( !pingSucceeded || healthResponse.getStatus() == ClusterHealthStatus.RED )
+            {
+                unregister();
+            }
+        }
+        catch ( Exception e )
+        {
+            unregister();
+
+            throw e;
+        }
+    }
+
+    @Deactivate
+    @SuppressWarnings("WeakerAccess")
+    public void deactivate()
+        throws IOException
+    {
+        activateExecutorService.shutdown();
+
+        unregister();
+
+        if ( client != null )
+        {
+            client.close();
+        }
+    }
+
+    private void register()
+    {
+        if ( this.clientReg != null )
+        {
+            return;
+        }
+
+        this.clientReg = context.registerService( EsClient.class, client, new Hashtable<>() );
```

Comment:
don't pass empty Hashtable. pass null. Then legacy class is not in use and it is a few nanoseconds faster :)